modules/wallet/files.go

```diff
@@ -22,18 +22,23 @@ type savedKey struct {
 
 // save writes the contents of a wallet to a file.
 func (w *Wallet) save() (err error) {
-	// convert key map to a slice
+	// Convert the key map to a slice.
 	keySlice := make([]savedKey, 0, len(w.keys))
 	for _, key := range w.keys {
 		keySlice = append(keySlice, savedKey{key.secretKey, key.unlockConditions})
 	}
 
-	// Write the data to file.
-	// TODO: Instead of using WriteFile, write to a temp file and then do a move.
-	err = ioutil.WriteFile(w.filename, encoding.Marshal(keySlice), 0666)
+	// Write the data to a temp file
+	err = ioutil.WriteFile(w.filename+"_temp", encoding.Marshal(keySlice), 0666)
 	if err != nil {
 		return
 	}
+	// Atomically overwrite the old wallet file with the new wallet file.
+	err = os.Rename(w.filename+"_temp", w.filename)
+	if err != nil {
+		// TODO: instruct user to recover wallet from w.filename+"_temp"
```

Comment:
The `err` here never indicates corruption, only that the move failed, which I believe only happens if 1) one or both paths do not exist, or 2) there's a permissions problem. Since we just wrote `w.filename`, 1) will only occur if it somehow gets deleted in the interim, in which case `w.filename` will still exist. Likewise, 2) should only occur if the permissions of `w.filename` have been changed since the last `save`, in which case it's up to the user to fix them.

Maybe for extra security we should have `load` try to load `w.filename+"_temp"` if loading `w.filename` fails?